Call impression URL when displaying sponsored homepage shelf

bobsilverberg opened the pull request:
~~Depends on https://github.com/mozilla/addons-frontend/pull/9758~~
Fixes #9720

willdurand commented on config/development.js:
```diff
@@ -32,6 +32,7 @@ module.exports = {
         "'self'",
         addonsServerDevCDN,
         analyticsHost,
+        apiDevHost,
```
mm why is this needed?

bobsilverberg replied:
I think you answered your own question below. It's needed to enable the `sendBeacon` for the new `/shelves/sponsored/impression/` endpoint when developing locally.

willdurand commented on src/amo/components/SponsoredAddonsShelf/index.js:
```diff
@@ -68,8 +83,37 @@ export class SponsoredAddonsShelfBase extends React.Component<InternalProps> {
+    if (_config.get('enableFeatureUseAdzerkForSponsoredShelf') && shelfData) {
+      const { impressionData, impressionURL } = shelfData;
+
+      if (impressionData && impressionURL) {
```
Should we log a debug message (at least) when we don't have data/url?

bobsilverberg replied:
Sure, good idea.

willdurand commented on src/amo/components/SponsoredAddonsShelf/index.js:
```diff
@@ -68,8 +83,37 @@ export class SponsoredAddonsShelfBase extends React.Component<InternalProps> {
+            data: impressionData,
+            key: 'impression_data',
+          }),
+          urlString: impressionURL,
```
I suppose the API is needed in the CSP because this url is absolute. I am wondering if that has any security implications. We could potentially make this URL relative but maybe it's fine as it is currently?

bobsilverberg replied:
We actually already include the apiHost in the `connectSrc` for all other environments. The only change in this patch is that we're adding it to the `development` config, so I don't think there are any security implications.

willdurand commented on tests/unit/amo/components/TestSponsoredAddonsShelf.js:
```diff
@@ -216,6 +262,82 @@ describe(__filename, () => {
       sinon.assert.notCalled(dispatchSpy);
     });
+
+    it('sends a ping for the impression on mount', () => {
```
is it "ping" or beacon?

bobsilverberg replied:
I'm calling it a ping as a colloquial term. You're right, it's a beacon. I'll change it.

willdurand commented on src/amo/components/SponsoredAddonsShelf/index.js:
```diff
@@ -68,8 +83,37 @@ export class SponsoredAddonsShelfBase extends React.Component<InternalProps> {
+    }
   }
 
+  sendImpressionPing = () => {
```
is it "ping" or "beacon"?

bobsilverberg replied:
I've changed this too.

willdurand commented on tests/unit/amo/utils/test_index.js:
```diff
@@ -259,4 +260,48 @@ describe(__filename, () => {
+      const _log = getFakeLogger();
+
+      sendBeacon({ _log, _navigator: null, urlString });
+      sinon.assert.calledWith(
```
Asserting that `sendBeacon` isn't actually called seems a bit better than asserting a log statement. I feel like we should not have to use `_log` in the function (and only use `log`).

bobsilverberg replied:
I would *love* to be able to assert that `sendBeacon` wasn't called, but the difficulty here is that we're testing a case where `_navigator` doesn't exist, which means that `_navigator.sendBeacon` is not a mock, and therefore there is no way for us to assert that. I agree that relying on log statements in a test is less than ideal, but this seemed like the best way to actually have a test for this, when we don't have a mock we can check.

willdurand replied:
![](https://media1.giphy.com/media/yvBAuESRTsETqNFlEl/giphy.gif)

bobsilverberg commented:
Thanks for the review, @willdurand. I am just about to land the dependency and then I will rebase this, which may require some changes. I'll reply to your comments after doing that.

bobsilverberg commented:
Ok, @willdurand, I have rebased it and addressed the issues you raised. I accidentally squashed the changes into the initial commit, because of the rebase. I hope that isn't too much of a hassle for you.

willdurand reviewed (approved):
r+wc

bobsilverberg commented:
Thanks @willdurand! Is this an r+wc because you wanted to make sure I saw your facepalm, or is there another comment that needs to be addressed before I can merge this?

willdurand commented:
I don't think there is any other comment
